### ISSUE TITLE
The problem with color balls in ASTCENC_PRE_VERYTHOROUGH

### ISSUE THREAD

**Issue by andrewevstyukhin** [ARM-software/astc-encoder#394]
Hi,
I meet color artifacts for relatively simple images.

Source image: 
![balls](https://user-images.githubusercontent.com/7923719/214050067-53bed44b-afaa-413b-b9b3-d1aefc7bf732.png)

:white_check_mark: Compressed image with 4x4 blocks is fine: 
![balls4x4](https://user-images.githubusercontent.com/7923719/214050186-cc0296cf-b6f3-40d6-88b3-ff72ed3506f0.png)

:exclamation: Compressed image with 5x5 blocks has problems: 
![balls5x5bug](https://user-images.githubusercontent.com/7923719/214050271-31a76f59-5f87-4233-99c6-6082f7edb6e9.png)

:exclamation: Compressed image with 6x6 blocks also has problems: 
![balls6x6bug](https://user-images.githubusercontent.com/7923719/214050378-f2931920-1fe9-4132-8480-afd5eb556739.png)

I use astcenc-4.2.0 (https://github.com/ARM-software/astc-encoder/releases/tag/4.2.0) with ASTCENC_PRE_VERYTHOROUGH preset and custom weights.

Imho it can be a bug in the logic.

**Comment by solidpixel** [comment]
> and custom weights.

... which are?

Try specifying the `-a 1` option to alpha-weight texel errors - it helps improve the alpha boundary, although doesn't completely remove block artefacts around the edges of the circles.

**Comment by andrewevstyukhin** [comment]
> config.cw_a_weight = 1.000f;
config.cw_g_weight = 0.715f;
config.cw_r_weight = 0.213f;
config.cw_b_weight = 0.072f;

For integer weights (those multiplied by 1000) I get messed up results for 4x4 blocks too.

**Comment by solidpixel** [comment]
Your extrudes look quite noisy compared the actual edge color. The dark pixels in the extruded region are forcing the compressor to select a much wider endpoint than is really needed.

Also I think your extrude borders can end up too small for the larger block sizes (e.g. 6x6), so are picking up either black pixels or neighbouring extrudes of a totally different color.

![extrudes](https://user-images.githubusercontent.com/23016617/214120540-a096f821-2499-4e02-8e05-be18a612f054.png)

**Comment by solidpixel** [comment]
Just to add one footnote on the comment above ...

> relatively simple images

Smooth circular gradients with a non-correlated alpha channel is one of the hardest things for a lossy compressor to encode!  They look simple, but they need a lot of precision in the encoding and the weight grid, and we often just run out of bits to encode them well. It is quite possible there isn't a "fix" for this one, but I'll see what I can improve.

**Comment by andrewevstyukhin** [comment]
✅ Just to brave up, please take a look at this ETC2 output:
![ballsEtc2](https://user-images.githubusercontent.com/7923719/214132606-8956a6bd-a0e0-4a3c-bd20-6992ca90d232.png)

I think ASTC is more powerful than ETC2, of course this is true for 4x4 blocks. I bet 5x5 blocks are good enough for 2D graphics.

✅ BC7 output is excellent:
![ballsBc7](https://user-images.githubusercontent.com/7923719/214133756-d2b37dfc-e74c-4dc9-b017-f7386608a111.png)

:zap: Finally PVRTC 4bpp output is: 
![ballsPvrtc](https://user-images.githubusercontent.com/7923719/214134513-daa29aea-fab9-48cf-840d-17c080f3e9f3.png)

**Comment by solidpixel** [comment]
> Just to brave up, please take a look at this ETC2 output:

Sure, but ETC2+EAC RGBA is 8bpp, which is the same as ASTC 4x4. 

> I bet 5x5 blocks are good enough for 2D graphics.

Yes, 5x5 seems "almost" good enough already - hopefully just some fine tuning. Suspect 6x6 will be a stretch though. 

Also, if you are able, don't rule out the non-square sizes for some intermediate bitrate options. 5x4 and 6x5 might also be worth trying.

**Comment by andrewevstyukhin** [comment]
Imho 6x6 is good for opaque textures.

It'll be very cute to achieve artifactless results for 5x5 blocks. Only exhaustive test on problematic regions can answer on our fundamental question.

**Comment by andrewevstyukhin** [comment]
❗ Compressed image with 5x4 blocks has minor problems:
![balls5x4bug](https://user-images.githubusercontent.com/7923719/214148938-93809425-41fb-4c0c-8de3-641e13352cc5.png)

**Comment by solidpixel** [comment]
Single 5x5 block reproducer of the solid color blocks around the edge:

Input:
![in-min](https://user-images.githubusercontent.com/23016617/214161176-50640d4b-6b5e-46d2-9f68-d0fbca7e7aec.png)


Output:
![out-min](https://user-images.githubusercontent.com/23016617/214161206-18089cec-dab1-406b-baa4-559d3e19bbb9.png)

**Comment by solidpixel** [comment]
One thing that does seem to help is relaxing the error weighting on the alpha channel, if you can tolerate slightly less accurate edges. Because alpha is full range due to the sharp transition, the alpha errors tend to dominate because they are high in magnitude compared to the comparatively small color variation.

```
astcenc-avx2 -tl in.png out.png 5x5 -verythorough -cw 1 1 1 0.5
```

![out-new](https://user-images.githubusercontent.com/23016617/214252556-d0e70ac4-61a1-46d2-8192-c554d529ba27.png)

**Comment by andrewevstyukhin** [comment]
Currently I use 4x4 blocks for the case. Anyway, 4x4 solution has wiped pixels:
![image](https://user-images.githubusercontent.com/7923719/214257358-f49ec0dd-4d95-422a-8266-7c7e820ce78a.png)

**Comment by andrewevstyukhin** [comment]
`astcenc-avx2.exe -tl balls.png balls4x4.png 4x4 -verythorough -yflip -cw 0.213 0.715 0.072 1.000`
has more glitches than
`astcenc-avx2.exe -tl balls.png balls4x4e.png 4x4 -exhaustive -yflip -cw 0.213 0.715 0.072 1.000`

What if exhaustive compression can be applied to badly packed blocks as a post-process step?

**Comment by andrewevstyukhin** [comment]
> (error1000) * (error1000) is a very big number

Do weights apply before squaring? If so, I urgently need square rooted weights.

I've already falled into the overflow because I changed float to integer weights at last moment and nobody noticed. I recently fixed it on my own when updated to excellent `astcenc-4.2.0`, which fixed blue-contraction 🐛.

**Comment by solidpixel** [comment]
> Do weights apply before squaring? 

Yes - it's a scale on the calculated error, and the compressor uses RMSE for ranking candidates, so weighting is applied before squaring. (Technically the compressor internally just uses MSE without rooting because the sqrt cannot change the relative ranking of results,).

The intent of channel weights is to adjust the relative balance of the channels, so "small floating point numbers" was always the intention of it. Using large weights everywhere is really just scaling the quality target, which is better achieved using the `-dblimit ` option.

**Comment by solidpixel** [comment]
> I noticed some discrepancy

Thanks, fixed on latest main branch.

**Comment by andrewevstyukhin** [comment]
I did some experiments under debugger with more kmeans steps and bigger TUNE_MAX_PARTITIIONING_CANDIDATES and TUNE_MAX_TRIAL_CANDIDATES values. `5x5 block reproducer` is still the same...

I have some ideas about kmeans which require more time: ASTC is a very complex format. First thing to try is a pixel mask about useful RGB values. I already have such mask to seamlessly combine classic RGBA and premultiplied alpha images into the same texture atlas. Technically I should zero RGB weights and errors for redundant pixels.

It may take in the month to find the clue. Currently choosing between `-verythorough` and `-exhaustive` is a solution.
I don't use exhaustive by default because of a long duration. Some test 2k.png timings on Skylake-X are:
> `DXT5` - SSE4.1 VEX, exhaustive alpha = 246 ms, refinement for RGB = 21 ms, `total 267 ms`
`ETC2_RGBA` - AVX2, fully exhaustive: alpha = 321 ms, RGB = 753 ms, `total 1074 ms`
`PVRTC 4bpp` - AVX2, refinement with 49x2 SVD, `total 2036 ms`
`PVRTC 2bpp` - AVX2, refinement with 49x2 SVD, `total 989 ms`
`BC7` - AVX512, nearly exhaustive with RDO and LAPACK SVD, `total 1873 ms`
`ASTC5x5 thorough` - AVX2, `total 2262 ms` perfect baseline
`ASTC5x5 verythorough` - AVX2, `total 7163 ms` maybe avx512 can help
`ASTC5x5 exhaustive` - AVX2, `total 21180 ms` too long

**Comment by solidpixel** [comment]
> I have some ideas about kmeans which require more time

I don't think this is a partition pattern selection problem - the blocks in question are relatively simple from a partitioning point of view (single diagonal split). Suspect the issue is more related to endpoint selection, amplified by the fast non-correlated alpha transition.

> I don't use exhaustive by default because of a long duration. 

Yes, compared to BC7 ASTC has a much much larger search space.

**Comment by andrewevstyukhin** [comment]
> Suspect the issue is more related to endpoint selection
Looks like a loose of extreme values (min & max). Also alpha weight can be adaptive to active color channels.

**Comment by solidpixel** [comment]
Current summary of the problem:

Alpha is getting treated preferentially by the encoder, because it changes from 1.0 to 0.0 fast (large range = large diffs = large errors) and the user weighting biases towards it more (in the original user channel weight settings). 

Could imagine changing to using relative errors vs the original min/max range, but I'm not sure how to weight that and have stable results, so I'd suggest just manually reducing the channel weight of alpha for now.
